src/OpenApi/perf/TransformersBenchmark.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection.Metadata;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Diagnosers;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi.Microbenchmarks;
+
+/// <summary>
+/// The following benchmarks are used to assess the memory and performance
+/// impact of different types of transformers. In particular, we want to
+/// measure the impact of (a) context-object creation and caching and (b)
+/// enumerator usage when processing operations in a given document.
+/// </summary>
+[MemoryDiagnoser]
+[EventPipeProfiler(EventPipeProfile.GcVerbose)]
+public class TransformersBenchmark : OpenApiDocumentServiceTestBase
+{
+    [Params(10, 100, 1000)]
+    public int InvocationCount { get; set; }
+
+    private readonly IEndpointRouteBuilder _builder = CreateBuilder();
+    private readonly OpenApiOptions _options = new OpenApiOptions();
+    private OpenApiDocumentService _documentService;
+
+    [GlobalSetup(Target = nameof(OperationTransformerAsDelegate))]
+    public void OperationTransformerAsDelegate_Setup()
+    {
+        _builder.MapGet("/", () => { });
+        for (var i = 0; i <= 1000; i++)
+        {
+            _options.UseOperationTransformer((operation, context, token) =>
+            {
+                operation.Description = "New Description";
+                return Task.CompletedTask;
+            });
+        }
+        _documentService = CreateDocumentService(_builder, _options);
+    }
+
+    [GlobalSetup(Target = nameof(ActivatedDocumentTransformer))]
+    public void ActivatedDocumentTransformer_Setup()
+    {
+        _builder.MapGet("/", () => { });
+        for (var i = 0; i <= 1000; i++)
+        {
+            _options.UseTransformer<ActivatedTransformer>();
+        }
+        _documentService = CreateDocumentService(_builder, _options);
+    }
+
+    [GlobalSetup(Target = nameof(DocumentTransformerAsDelegate))]
+    public void DocumentTransformerAsDelegate_Delegate()
+    {
+        _builder.MapGet("/", () => { });
+        for (var i = 0; i <= 1000; i++)
+        {
+            _options.UseTransformer((document, context, token) =>
+            {
+                document.Info.Description = "New Description";
+                return Task.CompletedTask;
+            });
+        }
+        _documentService = CreateDocumentService(_builder, _options);
+    }
+
+    [Benchmark]
```

Comment:
What were the results?  Are the comparable?

Reply:
Example results are below. The `OperationTransformerAsDelegate` is more allocate-y than the others because of the use of the enumerator over the `OpenApiOperations` dictionary. We could optimize this but it makes the code a little less easy to read so I avoided this for now.

| Method                         | InvocationCount | Mean         | Error      | StdDev     | Op/s      | Gen0     | Gen1   | Gen2   | Allocated   |
|------------------------------- |---------------- |-------------:|-----------:|-----------:|----------:|---------:|-------:|-------:|------------:|
| OperationTransformerAsDelegate | 10              |    502.24 us |   2.784 us |   2.604 us |  1,991.07 |   3.9063 | 0.9766 | 0.9766 |   436.21 KB |
| ActivatedDocumentTransformer   | 10              |     50.65 us |   0.510 us |   0.477 us | 19,744.86 |   0.3052 | 0.1221 | 0.0610 |    44.72 KB |
| DocumentTransformerAsDelegate  | 10              |    112.61 us |   0.952 us |   0.891 us |  8,879.85 |   0.3662 | 0.2441 | 0.1221 |    44.72 KB |
| OperationTransformerAsDelegate | 100             |  5,154.06 us |  99.184 us |  92.777 us |    194.02 |  31.2500 |      - |      - |  4282.05 KB |
| ActivatedDocumentTransformer   | 100             |    485.27 us |   3.190 us |   2.984 us |  2,060.69 |   2.9297 | 0.9766 |      - |   367.19 KB |
| DocumentTransformerAsDelegate  | 100             |  1,118.66 us |   4.202 us |   3.725 us |    893.93 |   1.9531 |      - |      - |   367.19 KB |
| OperationTransformerAsDelegate | 1000            | 49,849.30 us | 358.614 us | 299.459 us |     20.06 | 272.7273 |      - |      - | 42820.56 KB |
| ActivatedDocumentTransformer   | 1000            |  4,973.16 us |  61.507 us |  57.534 us |    201.08 |  23.4375 | 7.8125 |      - |  3671.91 KB |
| DocumentTransformerAsDelegate  | 1000            | 11,212.38 us |  81.451 us |  76.189 us |     89.19 |  15.6250 |      - |      - |  3671.91 KB |

Reply:
Don't we expect that one to be widely used?  I thought the whole point of exposing the helper was so that app developers wouldn't have to roll their own.  I think it might be worth a reduction in readability in this case.

Reply:
I included the optimization in https://github.com/dotnet/aspnetcore/commit/40a9912db7348010e67c2e20f0185dff8b4972cf.  Description of it is [here](https://github.com/dotnet/aspnetcore/commit/40a9912db7348010e67c2e20f0185dff8b4972cf#diff-ce401ad17ef25cda571313a54d84ecac390431805f07154f8cab3eed1e0756fcR12).

**Original (with Dictionary enumerator)**

| Method                         | TransformerCount | Mean      | Error     | StdDev    | Op/s      | Gen0   | Gen1   | Allocated |
|------------------------------- |----------------- |----------:|----------:|----------:|----------:|-------:|-------:|----------:|
| OperationTransformerAsDelegate | 10               |  2.156 us | 0.0176 us | 0.0147 us | 463,813.7 | 0.0343 | 0.0076 |   4.15 KB |
| ActivatedDocumentTransformer   | 10               |  1.686 us | 0.0216 us | 0.0202 us | 592,946.5 | 0.0324 | 0.0076 |   3.93 KB |
| DocumentTransformerAsDelegate  | 10               |  1.661 us | 0.0177 us | 0.0157 us | 601,973.3 | 0.0305 | 0.0076 |   3.67 KB |
| OperationTransformerAsDelegate | 100              |  7.310 us | 0.0464 us | 0.0387 us | 136,796.5 | 0.0610 | 0.0153 |   7.66 KB |
| ActivatedDocumentTransformer   | 100              |  3.801 us | 0.0311 us | 0.0276 us | 263,095.2 | 0.0496 | 0.0114 |   6.04 KB |
| DocumentTransformerAsDelegate  | 100              |  2.838 us | 0.0337 us | 0.0315 us | 352,392.1 | 0.0305 | 0.0076 |   3.67 KB |
| OperationTransformerAsDelegate | 1000             | 53.321 us | 0.1793 us | 0.1590 us |  18,754.3 | 0.3052 | 0.0610 |  42.82 KB |
| ActivatedDocumentTransformer   | 1000             | 32.117 us | 0.6423 us | 0.6872 us |  31,136.6 | 0.1831 |      - |  28.13 KB |
| DocumentTransformerAsDelegate  | 1000             | 12.615 us | 0.0257 us | 0.0201 us |  79,269.2 | 0.0153 |      - |   3.67 KB |

**Improved (with OperationType-lookup)**

| Method                         | TransformerCount | Mean      | Error     | StdDev    | Op/s      | Gen0   | Gen1   | Allocated |
|------------------------------- |----------------- |----------:|----------:|----------:|----------:|-------:|-------:|----------:|
| OperationTransformerAsDelegate | 10               |  2.561 us | 0.0293 us | 0.0274 us | 390,501.5 | 0.0305 | 0.0076 |    3.7 KB |
| ActivatedDocumentTransformer   | 10               |  1.686 us | 0.0108 us | 0.0096 us | 593,113.3 | 0.0324 | 0.0076 |   3.93 KB |
| DocumentTransformerAsDelegate  | 10               |  1.794 us | 0.0134 us | 0.0119 us | 557,380.3 | 0.0305 | 0.0076 |   3.67 KB |
| OperationTransformerAsDelegate | 100              |  8.255 us | 0.1338 us | 0.1252 us | 121,144.2 | 0.0153 |      - |    3.7 KB |
| ActivatedDocumentTransformer   | 100              |  4.016 us | 0.0770 us | 0.0824 us | 248,976.9 | 0.0458 | 0.0076 |   6.04 KB |
| DocumentTransformerAsDelegate  | 100              |  2.762 us | 0.0455 us | 0.0426 us | 362,068.0 | 0.0267 | 0.0076 |   3.67 KB |
| OperationTransformerAsDelegate | 1000             | 60.630 us | 0.1782 us | 0.1667 us |  16,493.6 |      - |      - |    3.7 KB |
| ActivatedDocumentTransformer   | 1000             | 35.310 us | 0.5062 us | 0.4735 us |  28,320.9 | 0.1831 |      - |  27.13 KB |
| DocumentTransformerAsDelegate  | 1000             | 12.969 us | 0.0904 us | 0.0755 us |  77,106.4 | 0.0153 |      - |   3.67 KB |